api/authenticator.py

```diff
@@ -26,162 +36,505 @@
+            )
+        else:
+            raise ValueError(
+                "Cannot create patron without some way of identifying them uniquely."
             )
-            return cls(DummyMilleniumPatronAPI(), [])
+        __transaction = _db.begin_nested()
+        patron, is_new = get_one_or_create(_db, Patron, **search_by)
+
+        # This makes sure the Patron is brought into sync with the
+        # other fields of this PatronData object, regardless of
+        # whether or not it is newly created.
+        if patron:
+            self.apply(patron)
+        __transaction.commit()
+        return patron, is_new
+
+    @property
+    def to_response_parameters(self):
+        """Return information about this patron which the client might
+        find useful.
+
+        This information will be sent to the client immediately after
+        a patron's credentials are verified by an OAuth provider.
+        """
+        if self.personal_name:
+            return dict(name=self.personal_name)
+        return None
 
-        providers = Configuration.policy("authentication")
-        if not providers:
+    def set_authorization_identifier(self, authorization_identifier):
+        """Helper method to set both .authorization_identifier
+        and .authorization_identifiers appropriately.
+        """
+        # The first authorization identifier in the list is the one
+        # we should use for Patron.authorization_identifier, assuming
+        # Patron.authorization_identifier needs to be updated.
+        if isinstance(authorization_identifier, list):
+            authorization_identifiers = authorization_identifier
+            authorization_identifier = authorization_identifiers[0]
+        elif authorization_identifier is None:
+            authorization_identifiers = []
+            authorization_identifier = None
+        elif authorization_identifier is self.NO_VALUE:
+            authorization_identifiers = []
+            authorization_identifier = self.NO_VALUE
+        else:
+            authorization_identifiers = [authorization_identifier]
+        self.authorization_identifier = authorization_identifier
+        self.authorization_identifiers = authorization_identifiers
+        
+class Authenticator(object):
+    """Use the registered AuthenticationProviders to turn incoming
+    credentials into Patron objects.
+    """    
+
+    @classmethod
+    def from_config(cls, _db):
+        """Initialize an Authenticator from site configuration.
+        """       
+        authentication_policy = Configuration.policy("authentication")
+        if not authentication_policy:
             raise CannotLoadConfiguration(
                 "No authentication policy given."
             )
+
+        if isinstance(authentication_policy, basestring):
+            authentication_policy = dict(providers=[authentication_policy])
+        bearer_token_signing_secret = authentication_policy.get(
+            'bearer_token_signing_secret'
```

Comment:
Since they started out as the same, should we keep them the same so that existing bearer tokens don't become invalid?